shopfloor_manual_product_transfer/services/manual_product_transfer.py

```diff
@@ -0,0 +1,668 @@
+        """Transition to the 'scan_product' state for the given source location."""
+        return self._response(
+            next_state="scan_product",
+            data={"location": self.data.location(location)},
+            message=message,
+        )
+
+    def _response_for_confirm_quantity(
+        self, location, product, quantity, lot=None, message=None
+    ):
+        """Transition to the 'confirm_quantity' state for the given move line."""
+        data = {
+            "location": self.data.location(location),
+            "product": self.data.product(product),
+            "quantity": quantity,
+        }
+        if lot:
+            data["lot"] = self.data.lot(lot)
+        return self._response(
+            next_state="confirm_quantity",
+            data=data,
+            message=message,
+        )
+
+    def _response_for_set_quantity(self, move_lines, quantity, message=None):
+        """Transition to the 'set_quantity' state for the given move line."""
+        return self._response(
+            next_state="set_quantity",
+            data={
+                "move_lines": self.data.move_lines(move_lines),
+                "quantity": quantity,
+            },
+            message=message,
+        )
+
+    def _response_for_scan_destination_location(
+        self, picking, move_lines, message=None
+    ):
+        """Transition to the 'scan_destination_location' state
+        for the given move lines.
+        """
+        return self._response(
+            next_state="scan_destination_location",
+            data={
+                "picking": self.data.picking(picking),
+                "move_lines": self.data.move_lines(move_lines),
+            },
+            message=message,
+        )
+
+    def scan_source_location(self, barcode):
+        """Scan a source location.
+
+        It is the starting point of this scenario.
+
+        If stock has been found in the scanned location, allows to scan a
+        product or a lot.
+
+        Transitions:
+        * scan_product: to scan a product or a lot stored in the scanned location
+        * start: no stock found or wrong barcode
+        """
+        search = self._actions_for("search")
+        location = search.location_from_scan(barcode)
+        if not location:
+            return self._response_for_start(message=self.msg_store.no_location_found())
+        if not self.is_src_location_valid(location):
+            return self._response_for_start(
+                message=self.msg_store.location_not_allowed()
+            )
+        quants = self.env["stock.quant"].search(
+            [("location_id", "=", location.id), ("quantity", ">", 0)]
+        )
+        if not quants:
+            return self._response_for_start(
+                message=self.msg_store.location_empty(location)
+            )
+        return self._response_for_scan_product(location)
+
+    def _find_user_move_lines_domain(self, location, product, lot=None):
```

Comment:
can't we reuse the search move line action?

Reply:
I checked the `_search_move_lines_by_location_domain` method, while some parts could be reused, we will have to delete/replace the following criteria from the returned domain:

- `("location_id", "child_of", locations.ids)`
- `("qty_done", "=", 0)`
- `("state", "in", ("assigned", "partially_available"))`
- `"|",
                ("shopfloor_user_id", "=", False),
                ("shopfloor_user_id", "=", self.env.uid),`

The only thing reusable at this stage is the domain on `product_id` and `lot_id`... so better to not reuse it IMO.
But I'll change the criteria `("state", "not in", ("cancel", "done"))`, I remember it makes things slower over time in the execution plan of PostgreSQL, and it's better to filter on `confirmed`, `assigned` and `partially_available` states as there are less records compared to `cancel`+`done`.